File: proxy.py
"""
Twilio Voice <-> Speechmatics Realtime proxy server (multi-channel).

Bridges Twilio Media Streams to Speechmatics' Realtime API using channel
diarization. This demo uses `track="both_tracks"`, so each call produces
two per-speaker transcript streams (Inbound / Outbound), transcribed
via a single Speechmatics RT multi-channel session.

Run:
    python proxy.py

The server listens on localhost:<PORT>/twilio and, on startup, opens
an ngrok tunnel and points the configured Twilio number's Voice URL at
the tunnel's /twiml endpoint. Dial the Twilio number to trigger a call.
"""

import asyncio
import base64
import json
import os
import sys
from http import HTTPStatus
from typing import Optional

import ngrok
import websockets
from dotenv import load_dotenv
from speechmatics.rt import (
    AsyncMultiChannelClient,
    AudioEncoding,
    AudioFormat,
    AuthenticationError,
    Model,
    ServerMessageType,
    TranscriptionConfig,
    TranscriptResult,
)
from twilio.rest import Client as TwilioClient
from twilio.twiml.voice_response import VoiceResponse

load_dotenv()

SM_API_KEY = os.getenv("SPEECHMATICS_API_KEY")
NGROK_AUTHTOKEN = os.getenv("NGROK_AUTHTOKEN")
TWILIO_ACCOUNT_SID = os.getenv("TWILIO_ACCOUNT_SID")
TWILIO_AUTH_TOKEN = os.getenv("TWILIO_AUTH_TOKEN")
TWILIO_PHONE_NUMBER = os.getenv("TWILIO_PHONE_NUMBER")
FORWARD_TO_NUMBER = os.getenv("FORWARD_TO_NUMBER")
PORT = int(os.getenv("SM_TWILIO_PROXY_PORT", "5000"))
CHANNELS = ["inbound", "outbound"]
LANGUAGE = "en"

ENV_VARS = (
    "SPEECHMATICS_API_KEY",
    "NGROK_AUTHTOKEN",
    "TWILIO_ACCOUNT_SID",
    "TWILIO_AUTH_TOKEN",
    "TWILIO_PHONE_NUMBER",
    "FORWARD_TO_NUMBER",
)

# Twilio Media Streams sends mu-law 8 kHz mono in 20ms frames (160 bytes each),
# base64-encoded inside JSON. Speechmatics Realtime accepts this natively.
TWILIO_AUDIO_FORMAT = AudioFormat(
    encoding=AudioEncoding.MULAW,
    sample_rate=8000,
    chunk_size=160,
)

class AsyncQueueSource:
    """File-like async source backed by an asyncio.Queue of audio frames.

    read() blocks until data is available, returning empty bytes only on
    explicit close() - otherwise the multi-channel SDK would treat a
    stall as EOF and tear down the whole session.
    """

    def __init__(self) -> None:
        self._queue: asyncio.Queue[bytes] = asyncio.Queue()
        self._buffer = b""
        self._closed = False

    def enqueue(self, data: bytes) -> None:
        if data and not self._closed:
            self._queue.put_nowait(data)

    def close(self) -> None:
        # Signal EOF; the sentinel unblocks any pending read().
        self._closed = True
        self._queue.put_nowait(b"")

    async def read(self, n: int) -> bytes:
        while len(self._buffer) < n:
            chunk = await self._queue.get()
            if not chunk:
                # EOF sentinel. Put it back so subsequent reads also see EOF,
                # then return whatever we have (may be a short read or empty).
                self._queue.put_nowait(b"")
                break
            self._buffer += chunk

        out, self._buffer = self._buffer[:n], self._buffer[n:]
        return out


class CallSession:
    """One Speechmatics multi-channel session covering both Twilio tracks."""

    def __init__(self) -> None:
        self.sources: dict[str, AsyncQueueSource] = {
            channel: AsyncQueueSource() for channel in CHANNELS
        }
        self._closed = False
        self._task: Optional[asyncio.Task] = None
        # Track the last speaker so consecutive finals from the same channel
        # continue on one line and speaker-changes start a new line.
        self._last_final_channel: Optional[str] = None

    def start(self) -> None:
        self._task = asyncio.create_task(self._run())

    def enqueue(self, track: str, frame: bytes) -> None:
        if self._closed or track not in self.sources:
            return
        self.sources[track].enqueue(frame)

    async def close(self) -> None:
        self._closed = True
        for source in self.sources.values():
            source.close()
        if self._task:
            await self._task

    async def _run(self) -> None:
        transcription_config = TranscriptionConfig(
            language=LANGUAGE,
            enable_partials=True,
            model=Model.ENHANCED,
            diarization="channel",
            channel_diarization_labels=CHANNELS,
        )

        try:
            async with AsyncMultiChannelClient(api_key=SM_API_KEY) as client:

                @client.on(ServerMessageType.ADD_TRANSCRIPT)
                def _on_final(msg):
                    text = TranscriptResult.from_message(msg).metadata.transcript
                    channel = msg.get("channel", "?")
                    if not text:
                        return
                    if self._last_final_channel is None:
                        sys.stdout.write(f"[{channel:<8}] {text}")
                    elif channel != self._last_final_channel:
                        sys.stdout.write(f"\n[{channel:<8}] {text}")
                    else:
                        sys.stdout.write(f" {text}")
                    self._last_final_channel = channel
                    sys.stdout.flush()

                await client.transcribe(
                    sources=self.sources,
                    transcription_config=transcription_config,
                    audio_format=TWILIO_AUDIO_FORMAT,
                )
        except AuthenticationError as e:
            print(f"Speechmatics auth error: {e}", file=sys.stderr)
        except Exception as e:  # noqa: BLE001
            if not self._closed:
                print(
                    f"Speechmatics session error: {type(e).__name__}: {e}",
                    file=sys.stderr,
                )
        finally:
            # Stop accepting further audio from Twilio; the RT connection
            # already closed on the way out of the `async with` above.
            self._closed = True
            for source in self.sources.values():
                source.close()


async def handle_twilio_call(ws) -> None:
    """Consume a Twilio Media Streams WebSocket and pump audio into a session."""
    session: Optional[CallSession] = None
    stream_sid: Optional[str] = None

    try:
        async for raw in ws:
            message = json.loads(raw)
            event = message.get("event")

            if event == "connected":
                print("[twilio] protocol handshake ok")

            elif event == "start":
                start = message["start"]
                stream_sid = start.get("streamSid")
                print(
                    f"[twilio] call started - stream={stream_sid} "
                    f"call={start.get('callSid')} tracks={start.get('tracks')}"
                )
                session = CallSession()
                session.start()

            elif event == "media" and session is not None:
                media = message["media"]
                # Twilio's runtime `media.track` is always bare "inbound"/"outbound"
                payload = media.get("payload")
                if payload:
                    session.enqueue(media.get("track", "inbound"), base64.b64decode(payload))

            elif event == "stop":
                print(f"[twilio] call stopped (stream={stream_sid})")
                break

    except websockets.exceptions.ConnectionClosed:
        print(f"[twilio] connection closed (stream={stream_sid})")
    finally:
        if session:
            await session.close()


async def ws_handler(websocket, path):
    if path == "/twilio":
        print("[twilio] incoming Media Streams connection")
        await handle_twilio_call(websocket)
    else:
        print(f"[twilio] ignoring connection on path {path}")


def _check_env_vars() -> None:
    missing = [name for name in ENV_VARS if not os.getenv(name)]
    if missing:
        print(f"Missing required env vars in .env: {', '.join(missing)}", file=sys.stderr)
        sys.exit(1)


async def _start_ngrok_tunnel() -> str:
    """Returns the tunnel base URL (https://...)."""
    listener = await ngrok.forward(addr=PORT, authtoken=NGROK_AUTHTOKEN, proto="http")
    return listener.url()


def _build_twiml(stream_url: str, forward_to: str) -> bytes:
    """Build the TwiML Twilio will fetch when a call comes in."""
    response = VoiceResponse()
    start = response.start()
    start.stream(url=stream_url, track="both_tracks")
    response.say("This call is being transcribed.", voice="woman", language="en")
    response.dial(forward_to)
    return str(response).encode()


async def _configure_twilio_number(twiml_url: str) -> None:
    """Point the Twilio number's voice URL at our /twiml endpoint."""
    client = TwilioClient(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN)
    numbers = await asyncio.to_thread(
        client.incoming_phone_numbers.list,
        phone_number=TWILIO_PHONE_NUMBER,
    )
    number = next((n for n in numbers if n.phone_number == TWILIO_PHONE_NUMBER), None)
    if number is None:
        print(f"Twilio number {TWILIO_PHONE_NUMBER} not found on this account.", file=sys.stderr)
        sys.exit(1)
    await asyncio.to_thread(
        client.incoming_phone_numbers(number.sid).update,
        voice_url=twiml_url,
        voice_method="GET",
    )


async def main() -> None:
    _check_env_vars()

    base_url = await _start_ngrok_tunnel()
    wss_url = base_url.replace("https://", "wss://") + "/twilio"
    twiml_body = _build_twiml(wss_url, FORWARD_TO_NUMBER)
    await _configure_twilio_number(base_url + "/twiml")

    async def _process_request(path, _headers):
        if path.split("?", 1)[0] == "/twiml":
            return (
                HTTPStatus.OK,
                [
                    ("Content-Type", "application/xml"),
                    ("Content-Length", str(len(twiml_body))),
                ],
                twiml_body,
            )
        return None

    async with websockets.serve(
        ws_handler, "localhost", PORT, process_request=_process_request
    ):
        print(f"\nListening on ws://localhost:{PORT}/twilio")
        print(f"Twilio number {TWILIO_PHONE_NUMBER} pointed at {wss_url}. Call it now.\n")
        await asyncio.Future()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nShutting down.")

File: js/proxy.ts
/**
 * Twilio Voice <-> Speechmatics Realtime proxy server (multi-channel).
 *
 * Node.js/TypeScript sibling of ../proxy.py. Bridges Twilio Media Streams to
 * Speechmatics' Realtime API using channel diarization: each call produces
 * two per-speaker transcript streams (inbound / outbound), transcribed via
 * a single Speechmatics RT multi-channel session.
 *
 * Run:
 *     pnpm start          (or `npm start`)
 *
 * The server listens on localhost:<PORT>/twilio and, on startup, opens an
 * ngrok tunnel and points the configured Twilio number's Voice URL at the
 * tunnel's /twiml endpoint. Dial the Twilio number to trigger a call.
 *
 * Environment variables (loaded from ../.env):
 *   SPEECHMATICS_API_KEY   Speechmatics Realtime API key.
 *   NGROK_AUTHTOKEN        ngrok auth token (for the public tunnel).
 *   TWILIO_ACCOUNT_SID     Twilio Account SID.
 *   TWILIO_AUTH_TOKEN      Twilio Auth Token.
 *   TWILIO_PHONE_NUMBER    E.164 number to auto-configure.
 *   FORWARD_TO_NUMBER      E.164 number the <Dial> bridges to.
 *   SM_TWILIO_PROXY_PORT   Optional. Local server port. Default: 5000.
 */

import http from "node:http";
import { fileURLToPath } from "node:url";
import path from "node:path";

import dotenv from "dotenv";
import ngrok from "@ngrok/ngrok";
import { WebSocketServer, type WebSocket } from "ws";
import { createSpeechmaticsJWT } from "@speechmatics/auth";
import {
  RealtimeClient,
  type AddChannelAudio,
  type EndOfChannel,
  type RealtimeClientMessage,
  type RealtimeTranscriptionConfig,
  type StartRecognition,
} from "@speechmatics/real-time-client";
import twilio from "twilio";

// Load ../.env so JS and Python demos can share the same secrets file.
const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, "..", ".env") });

// ─────────────────────────────────────────────────────────────────────
// Configuration
// ─────────────────────────────────────────────────────────────────────

const SM_API_KEY = process.env.SPEECHMATICS_API_KEY;
const NGROK_AUTHTOKEN = process.env.NGROK_AUTHTOKEN;
const TWILIO_ACCOUNT_SID = process.env.TWILIO_ACCOUNT_SID;
const TWILIO_AUTH_TOKEN = process.env.TWILIO_AUTH_TOKEN;
const TWILIO_PHONE_NUMBER = process.env.TWILIO_PHONE_NUMBER;
const FORWARD_TO_NUMBER = process.env.FORWARD_TO_NUMBER;
const PORT = Number.parseInt(process.env.SM_TWILIO_PROXY_PORT ?? "5000", 10);

const REQUIRED_ENV_VARS = [
  "SPEECHMATICS_API_KEY",
  "NGROK_AUTHTOKEN",
  "TWILIO_ACCOUNT_SID",
  "TWILIO_AUTH_TOKEN",
  "TWILIO_PHONE_NUMBER",
  "FORWARD_TO_NUMBER",
] as const;

// Twilio Media Streams sends mulaw 8 kHz mono in 20 ms frames (160 bytes
// each), base64-encoded inside JSON. Speechmatics Realtime accepts this
// natively - no transcoding.
const AUDIO_FORMAT = {
  type: "raw",
  encoding: "mulaw",
  sample_rate: 8000,
} as const;

// The channel IDs we register with Speechmatics match Twilio's own track
// labels so routing is a direct lookup - no translation table needed.
export const CHANNELS = ["inbound", "outbound"] as const;
export type Channel = (typeof CHANNELS)[number];

const LANGUAGE = "en";

// ─────────────────────────────────────────────────────────────────────
// Env validation
// ─────────────────────────────────────────────────────────────────────

function checkEnvVars(): void {
  const missing = REQUIRED_ENV_VARS.filter((name) => !process.env[name]);
  if (missing.length > 0) {
    console.error(`Missing required env vars in .env: ${missing.join(", ")}`);
    process.exit(1);
  }
}

// ─────────────────────────────────────────────────────────────────────
// MultiChannelRealtimeClient
//
// The JS SDK's RealtimeClient exposes sendAudio(binary) for single-channel
// raw audio, but no public sendChannelAudio(channel, base64). The
// multi-channel protocol *is* fully supported by the server - we just have
// to construct AddChannelAudio JSON messages ourselves. The SDK's own
// sendMessage() is private, so this subclass provides thin, typed helpers.
// ─────────────────────────────────────────────────────────────────────

interface PrivateSend {
  sendMessage(msg: RealtimeClientMessage): void;
}

class MultiChannelRealtimeClient extends RealtimeClient {
  sendChannelAudio(channel: string, base64Data: string): void {
    const msg: AddChannelAudio = {
      message: "AddChannelAudio",
      channel,
      data: base64Data,
    };
    (this as unknown as PrivateSend).sendMessage(msg);
  }

  endChannel(channel: string, lastSeqNo: number): void {
    const msg: EndOfChannel = {
      message: "EndOfChannel",
      channel,
      last_seq_no: lastSeqNo,
    };
    (this as unknown as PrivateSend).sendMessage(msg);
  }
}

// ─────────────────────────────────────────────────────────────────────
// CallSession - one Speechmatics session covering both Twilio tracks
// ─────────────────────────────────────────────────────────────────────

export class CallSession {
  private client: MultiChannelRealtimeClient | null = null;
  private closed = false;
  private lastFinalChannel: string | null = null;
  private readonly seqNo: Record<Channel, number> = { inbound: 0, outbound: 0 };
  private startPromise: Promise<void> | null = null;
  private eotResolve: (() => void) | null = null;
  private readonly eotPromise: Promise<void>;

  constructor() {
    // Track EndOfTranscript arrival so close() can wait for the server
    // to flush any final results before disconnecting.
    this.eotPromise = new Promise<void>((resolve) => {
      this.eotResolve = resolve;
    });
  }

  /** Kick off the Speechmatics RT connection. Non-blocking. */
  start(): void {
    this.startPromise = this.run();
  }

  /** Route a base64-encoded audio frame from Twilio to Speechmatics. */
  enqueue(track: string, base64Payload: string): void {
    if (this.closed || !this.isKnownChannel(track) || !this.client) {
      return;
    }
    this.seqNo[track] += 1;
    this.client.sendChannelAudio(track, base64Payload);
  }

  /**
   * Cleanly shut down the RT session: EndOfChannel per track, wait for
   * EndOfTranscript so any last finals print, then disconnect.
   */
  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;

    if (this.client) {
      for (const channel of CHANNELS) {
        try {
          this.client.endChannel(channel, this.seqNo[channel]);
        } catch {
          // Socket may already be closed - non-fatal.
        }
      }

      // Wait up to 10s for the server to flush the last transcripts
      // (EndOfTranscript arrives after all EndOfChannel are ack'd).
      await Promise.race([
        this.eotPromise,
        new Promise<void>((resolve) => setTimeout(resolve, 10_000)),
      ]);

      try {
        await this.client.stopRecognition({ noTimeout: true });
      } catch {
        // Already-closed sockets throw; safe to swallow on teardown.
      }

      process.stdout.write("\n");
    }

    if (this.startPromise) {
      // Ensure the run task settles before returning.
      await this.startPromise.catch(() => undefined);
    }
  }

  private isKnownChannel(track: string): track is Channel {
    return (CHANNELS as readonly string[]).includes(track);
  }

  private async run(): Promise<void> {
    if (!SM_API_KEY) throw new Error("SPEECHMATICS_API_KEY missing at runtime");

    const jwt = await createSpeechmaticsJWT({
      type: "rt",
      apiKey: SM_API_KEY,
      ttl: 60,
    });

    this.client = new MultiChannelRealtimeClient();

    this.client.addEventListener("receiveMessage", ({ data }) => {
      if (data.message === "EndOfTranscript") {
        this.eotResolve?.();
        return;
      }
      if (data.message === "AddPartialTranscript") {
        // Stub - wire partials here if you want to render them (live UI,
        // LLM prompt, agent-assist, etc.). Example:
        //   const text = data.results.map(r => r.alternatives?.[0].content).join(" ");
        //   const channel = data.channel ?? "?";
        return;
      }
      if (data.message === "AddTranscript") {
        this.handleFinal(data);
        return;
      }
      if (data.message === "Error") {
        console.error(`Speechmatics error: ${data.type} - ${data.reason ?? ""}`);
      }
    });

    const startMsg: RealtimeTranscriptionConfig = {
      transcription_config: {
        language: LANGUAGE,
        max_delay: 0.7,
        enable_partials: true,
        operating_point: "enhanced",
        diarization: "channel",
        channel_diarization_labels: [...CHANNELS],
      },
      audio_format: AUDIO_FORMAT,
    };

    try {
      await this.client.start(jwt, startMsg);
    } catch (err) {
      if (!this.closed) {
        console.error(
          `Speechmatics session error: ${err instanceof Error ? err.message : String(err)}`,
        );
      }
    }
  }

  private handleFinal(msg: {
    results: Array<{ alternatives?: Array<{ content: string }> }>;
    channel?: string;
  }): void {
    const text = msg.results.map((r) => r.alternatives?.[0]?.content ?? "").join(" ");
    if (!text) return;

    const channel = msg.channel ?? "?";
    const label = `[${channel.padEnd(8)}]`;

    if (this.lastFinalChannel === null) {
      process.stdout.write(`${label} ${text}`);
    } else if (channel !== this.lastFinalChannel) {
      process.stdout.write(`\n${label} ${text}`);
    } else {
      process.stdout.write(` ${text}`);
    }
    this.lastFinalChannel = channel;
  }
}

// ─────────────────────────────────────────────────────────────────────
// Twilio Media Streams WebSocket handler
// ─────────────────────────────────────────────────────────────────────

interface TwilioMediaEvent {
  event: string;
  start?: {
    streamSid?: string;
    callSid?: string;
    tracks?: string[];
  };
  media?: {
    track?: string;
    payload?: string;
  };
}

export async function handleTwilioCall(ws: WebSocket): Promise<void> {
  // Wrap the session in an object so TypeScript's control-flow analysis
  // doesn't wrongly infer "never assigned" from the outer scope - reads
  // via `.session` correctly narrow after the callback mutates it.
  const state: { session: CallSession | null; streamSid?: string } = {
    session: null,
  };

  const closed = new Promise<void>((resolve) => {
    ws.on("close", () => resolve());
    ws.on("error", () => resolve());
  });

  ws.on("message", (raw: Buffer) => {
    let message: TwilioMediaEvent;
    try {
      message = JSON.parse(raw.toString("utf8"));
    } catch {
      return;
    }

    switch (message.event) {
      case "connected":
        console.log("[twilio] protocol handshake ok");
        break;
      case "start": {
        state.streamSid = message.start?.streamSid;
        console.log(
          `[twilio] call started - stream=${state.streamSid} call=${message.start?.callSid} tracks=${JSON.stringify(message.start?.tracks)}`,
        );
        state.session = new CallSession();
        state.session.start();
        break;
      }
      case "media": {
        if (!state.session || !message.media?.payload) return;
        // Twilio's `media.track` is always bare "inbound"/"outbound" -
        // the "_track" suffix only appears in the TwiML attribute.
        const track = message.media.track ?? "inbound";
        // Base64 pass-through: Twilio's payload feeds AddChannelAudio.data
        // as-is, no decode/re-encode needed.
        state.session.enqueue(track, message.media.payload);
        break;
      }
      case "stop":
        console.log(`[twilio] call stopped (stream=${state.streamSid})`);
        ws.close();
        break;
    }
  });

  await closed;
  if (state.session) {
    await state.session.close();
  }
}

// ─────────────────────────────────────────────────────────────────────
// ngrok tunnel + TwiML + Twilio number configuration
// ─────────────────────────────────────────────────────────────────────

async function startNgrokTunnel(): Promise<string> {
  const listener = await ngrok.forward({
    addr: PORT,
    authtoken: NGROK_AUTHTOKEN,
    proto: "http",
  });
  const url = listener.url();
  if (!url) throw new Error("ngrok tunnel returned no URL");
  return url;
}

export function buildTwiml(streamUrl: string, forwardTo: string): string {
  const response = new twilio.twiml.VoiceResponse();
  const start = response.start();
  start.stream({ url: streamUrl, track: "both_tracks" });
  response.say(
    { voice: "woman", language: "en-GB" },
    "This call is being transcribed.",
  );
  response.dial(forwardTo);
  return response.toString();
}

async function configureTwilioNumber(twimlUrl: string): Promise<void> {
  const client = twilio(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN);
  // The `phoneNumber` filter is a substring/wildcard match; verify locally
  // that we pick the exact E.164 the user configured.
  const numbers = await client.incomingPhoneNumbers.list({
    phoneNumber: TWILIO_PHONE_NUMBER,
  });
  const match = numbers.find((n) => n.phoneNumber === TWILIO_PHONE_NUMBER);
  if (!match) {
    console.error(`Twilio number ${TWILIO_PHONE_NUMBER} not found on this account.`);
    process.exit(1);
  }
  await client.incomingPhoneNumbers(match.sid).update({
    voiceUrl: twimlUrl,
    voiceMethod: "GET",
  });
}

// ─────────────────────────────────────────────────────────────────────
// HTTP + WS server
// ─────────────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  checkEnvVars();

  const baseUrl = await startNgrokTunnel();
  const wssUrl = baseUrl.replace(/^https:/, "wss:") + "/twilio";
  const twimlBody = buildTwiml(wssUrl, FORWARD_TO_NUMBER as string);
  await configureTwilioNumber(baseUrl + "/twiml");

  const httpServer = http.createServer((req, res) => {
    const url = req.url ?? "/";
    if (url.split("?", 1)[0] === "/twiml") {
      res.writeHead(200, {
        "Content-Type": "application/xml",
        "Content-Length": Buffer.byteLength(twimlBody).toString(),
      });
      res.end(twimlBody);
      return;
    }
    res.writeHead(404);
    res.end();
  });

  const wss = new WebSocketServer({ noServer: true });
  httpServer.on("upgrade", (request, socket, head) => {
    const url = request.url ?? "";
    if (url.split("?", 1)[0] !== "/twilio") {
      socket.destroy();
      return;
    }
    wss.handleUpgrade(request, socket, head, (ws) => {
      console.log("[twilio] incoming Media Streams connection");
      handleTwilioCall(ws).catch((err) => {
        console.error(`Handler error: ${err instanceof Error ? err.message : String(err)}`);
      });
    });
  });

  await new Promise<void>((resolve) => {
    httpServer.listen(PORT, "localhost", () => resolve());
  });

  console.log(`\nListening on ws://localhost:${PORT}/twilio`);
  console.log(
    `Twilio number ${TWILIO_PHONE_NUMBER} pointed at ${wssUrl}. Call it now.\n`,
  );

  // Block forever until SIGINT.
  await new Promise<void>((resolve) => {
    process.on("SIGINT", () => {
      console.log("\nShutting down.");
      resolve();
    });
  });

  await new Promise<void>((resolve) => httpServer.close(() => resolve()));
}

// Only run main() when this file is executed directly (not when imported
// by tests).
const invokedDirectly =
  process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url);

if (invokedDirectly) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}
